Ergonomic changes for I2S callbacks.

dacut opened the pull request:
This makes a few changes suggested in the last bit of https://github.com/esp-rs/esp-idf-hal/pull/232:

* Trait functions take impl FnMut instead of Box<dyn FnMut>.
* Callback bookkeeping is no longer static (like the GPIO callbacks); instead, it's stored with the channel.
* Added rx/tx_unsubscribe() methods to the channels.

Also, fixed the naming of the I2sRxEvent/I2sTxEvent enums to be more inline with the brevity of Rust in general:
* DataReceived -> RxDone
* ReceiveOverflow -> RxOverflow
* DataTransmitted -> TxDone
* TransmitOverflow -> TxOverflow

ivmarkov commented:
@dacut: Overall, LGTM, two issues only:
* Can you look at the CI build failures? The CI pipeline is now functioning OK, so these are real issues.
* I think you are leaking the `Box` that you push into `UnsafeCallback`. The `UnsafeCallback::callback` member specifically. You need a `Drop` impl for `UnsafeCallback` that does `Box::from_raw` on the `callback` member and then does nothing with it so that the memory of the `Box` is dropped at the end of the `drop()` fn.

dacut commented:
Ah, good catch on the leak. I forgot to add `Drop` when I changed `Box<Box<...>>` to `*mut Box<...>`.

I think I've fixed the no-alloc bit as well (and beat my CI pipeline into submission as well based on your fixes); looks better now: https://github.com/dacut/esp-idf-hal/actions/runs/5380466584

ivmarkov commented:
Thanks!

ivmarkov commented:
@dacut Question: are you using the Callback API for anything? The one use case I find for it is `async` send/receive as I mentioned earlier. However, if that's the only use case we can do with the callbacks, it can be implemented in a simpler way, **without** - actually - exposing an unsafe `subscribe`/`unsubscribe` API tot he end user. The callbacks would just become an internal implementation detail of two new methods - `send_async` and `receive_async`.

Asking because if we take that route, I need to remove the Callback API and replace it with the above two async methods. The benefit: simpler implementation, and the demotion of the callback API would actually remove the `[cfg(feature = "alloc")]` stuff from the codebase, as implementing callbacks specifically for `async` support does not need boxing.

What do you think?

BTW, this question and what I'm suggestion has consequences beyond I2S. I'm currently working on exposing the continuous, DMA-powered ADC driver of ESP IDF, and I face the same situation there w.r.t. callbacks. I wouldn't be surprised if it is the same also for SPI, UART, I2C and so on.

dacut commented:
@ivmarkov I am not. The main use case I envisioned for it was processing DMA buffers in a just-in-time manner. I think `async` is a good fit here, but I have not played with async code on ESP32 yet (that usually means bringing in Tokio to me in the non-embedded world, but here I'm guessing it just means plugging the async subsystem into FreeRTOS).

I do wish the ESP-IDF SDK exposed the DMA buffers directly, though. It ends up [copying the data over](https://github.com/espressif/esp-idf/blob/master/components/driver/i2s/i2s_common.c#L1137), which I found a bit limiting on performance (though I was trying crazy things like spectrum analysis/FFTs).

Also, the ESP32 (but no other models) supports a direct I2S to ADC/DAC mode that isn't exposed in the ESP-IDF SDK (similar to the LCD/Camera mode). Let me know if that's of interest to you. (I'd have to toy a bit with these modes; I wish they had been kept in later models as I find them useful, but at least LCD mode is buggy on the ESP32 so I suspect that factored into the decision.)

ivmarkov commented:
> @ivmarkov I am not. The main use case I envisioned for it was processing DMA buffers in a just-in-time manner. I think `async` is a good fit here, but I have not played with async code on ESP32 yet (that usually means bringing in Tokio to me in the non-embedded world, but here I'm guessing it just means plugging the async subsystem into FreeRTOS).

OK, good to know. Then in the name of simplicity, I might be removing the `subscribe``unsubscribe` API and replacing it with a simple hidden notification mechanism to power the future async send/receive pair of functions.

W.r.t. `tokio` - no, you don't need that. I'm personally using a tailoring of `async-task` [here](https://github.com/ivmarkov/edge-executor). In future, we might be supporting `async-executor` as well, from the `embassy` ecosystem (though it needs all futures to be ultimately resolving to functions which take non-generic parameters, which is a bit of a restriction).

The nice thing about `async-task` (aside from it being minimal and amenable to customization) is that it plays well with `async-io`, which we can natively support on the ESP32 (unlike `tokio`'s `mio` which requires Linux-style `epoll` which we cannot support on the ESP IDF). `async-io` gives you `async` for network sockets.


> I do wish the ESP-IDF SDK exposed the DMA buffers directly, though. It ends up [copying the data over](https://github.com/espressif/esp-idf/blob/master/components/driver/i2s/i2s_common.c#L1137), which I found a bit limiting on performance (though I was trying crazy things like spectrum analysis/FFTs).

I think zero-copy DMA might be difficult with the `async` metaphor in Rust, but I could be wrong.

> Also, the ESP32 (but no other models) supports a direct I2S to ADC/DAC mode that isn't exposed in the ESP-IDF SDK (similar to the LCD/Camera mode). Let me know if that's of interest to you. (I'd have to toy a bit with these modes; I wish they had been kept in later models as I find them useful, but at least LCD mode is buggy on the ESP32 so I suspect that factored into the decision.)

We might be looking into this in future. For now, for me personally at least - I2S IO and continuous ADC is good enough.

dacut commented:
Ah, yeah, wasn't proposing bringing in Tokio here. It's just my sole exposure to async Rust for now; I'll take a look at what you're doing with `async-task`.

I'd have to think how DMA and async work together a bit. The read API is easy enough: `async fn read_async(self: Pin<&mut Self>, data: &[u8])`, where data is created via `slice::from_raw_parts`.

The write API is trickier, because you now need _two_ APIs: one to get the next buffer, one to submit it:
* `async fn wait_write_buffer(self: Pin<&mut Self>) -> &mut [MaybeUninit<u8>]`
* `async fn write_async(self: Pin<&mut Self>, data: &[u8])`

But this is insufficient:
* You need to make sure data is the same slice returned from `wait_write_buffer`.
* If the caller never calls `write_async`, there's likely cleanup work that needs to be done.

This calls for a struct to wrap the buffer, which I need more time to think about.

ivmarkov commented:
> I'd have to think how DMA and async work together a bit. The read API is easy enough: `async fn read_async(self: Pin<&mut Self>, data: &[u8])`, where data is created via `slice::from_raw_parts`.
> 
> The write API is trickier, because you now need _two_ APIs: one to get the next buffer, one to submit it:
> 
> * `async fn wait_write_buffer(self: Pin<&mut Self>) -> &mut [MaybeUninit<u8>]`
> * `async fn write_async(self: Pin<&mut Self>, data: &[u8])`

IMO neither `read` nor `write` is easy. Just look at the `Embassy` Matrix channel. Zero-copy DMA is like *the* topic for the past year, if not more. :)

My (very superficial) understanding of the zero-copy DMA problem is that is is roughly isomorphic to the Linux `io_uring` problem. In other words, both `io_uring` and zero-copy DMA are essentially *completion based*, while the Rust `Future` model is polling based. Intuitively, a completion based model means you need to "hand off" ownership of your read or write buffer to the DMA engine for the duration of the read/write operation. Once the DMA operation is over (and you are notified via a callback), you can reclaim the ownership of your buffer, but what you should NOT do - by any means - is to drop/free the buffer while the DMA operation is running. The problem is, typical `& [u8]` and `&mut [u8]` buffers are only valid while the Rust future is alive. But since you can `drop` the future at any time, these references might become invalid while the DMA engine is still using them, which is the crux of the problem. Now, if you can - in your `Future::drop` - stop the DMA engine in a quick, synchronous way, you still have a way out (kinda), as this way you reclaim your buffers "in emergency". This is still half-baked though, as if you just call `core::mem::forget(my_future)` (which is a completely safe Rust API), the `drop` won't be called, yet the lifetime of the `& [u8]` / `&mut [u8]` reference would be over, so you can move / drop your buffers immediately after `forget`-ing the future and still mess it up completely, as the DMA engine might still be accessing that memory!

You can of course refcount the buffers - thus allowing shared ownership - but if you do this, you take another hit - you need to allocate the memory from the heap so the thing is not zero cost for other reasons now. Or you just let the DMA engine have its own buffers, and then you copy from/to those buffers, as the ESP IDF driver does.

There has been a lot of talks in the community how to remedy the situation, including linear types, async drop and whatnot, but in the meantime, supporting a completion-based API from Rust means you have to choose your poison: refcounting and a heap allocation, or non-zero-copy.
